# **Лекция 4. Простейшие системы массового обслуживания и их характеристики**

- Одноканальная СМО с отказом
- Многоканальная СМО с отказом
- Одноканальная СМО с неограниченной очередью
- Многоканальная СМО с неограниченной очередью
- Одноканальная СМО с ограничением на длину очереди
- Многоканальная СМО с ограничением на длину очереди
- Описание пакета моделирования СМО queueing в системе R
- Задачи для самостоятельного решения

## $\S 1.$ **Одноканальная СМО с отказом $M/M/1/1$**

Классическим примером одноканальной системы с отказом в обслуживании может служить сокетное соединение с одним подключением. В тот момент, когда сокет занят взаимодействием с клиентом, остальные клиенты, пришедшие на обслуживание получают отказ, т.е. образование очереди в такой системе невозможно.

Пусть $\lambda, \mu$ - интенсивности потока заявок и обслуживания соответственно. Состояние системы $S_0$ обозначает, что канал свободен, $S_1$  означает, что канал занят обслуживанием.

<center><img src="1.png" width=300 height=300></center>

Составляя уравнения Колмогорова для стационарного режима, получаем:
$$\begin{cases} \lambda\cdot P_0=\mu\cdot P_1 \\ \mu\cdot P_1=\lambda\cdot P_0 \end{cases}$$
Таким образом, система вырождается в одно уравнение. Учитывая условие нормировки $P_0+P_1=1$, получаем:
$$P_0=\frac{\mu}{\lambda+\mu}; P_1=\frac{\lambda}{\lambda+\mu}\tag{1.1}$$

Полученные значения вероятностей $P_0, P_1$ по сути определяют длительность нахождения системы в состояниях $S_0, S_1$ соответственно. Вероятность нахождения системы в состоянии $S_1$ соответствует вероятности отказа: $P_{\text{отк}}=P_1$.

**Абсолютная пропускная способность** $$\lambda'=(1-P_1)\lambda=\frac{\lambda\mu}{\lambda+\mu}\tag{1.2}$$

**Относительная пропускная способность** (или вероятность обслуживания поступившей заявки):
$$Q=\frac{\lambda'}{\lambda} = P_0\tag{1.3}$$

**Интенсивность потока потерянных заявок**: 
$$\lambda''=P_{\text{отк}}\lambda=\frac{\lambda^2}{\lambda+\mu}\tag{1.4}$$

## $\S 2.$ **Многоканальная СМО с отказами $M/M/n/n$**

Рассмотрим одну из первых "классических" задач теории массового обслуживания. Эта задача возникла из практических нужд телефонии и была решена Эрлангом.

Задача ставится так:

*Имеется $n$ каналов, на которые поступает поток заявок с интенсивностью $\lambda$. Поток обслуживания одним каналом имеет интенсивность $\mu$. Найти финальные вероятности состояний СМО, а также характеристики эффективности:* 
- абсолютную пропускную способность (среднее число заявок, обслуживаемых в единицу времени);
- вероятность отказа;
- среднее число занятых каналов.

Введем обозначения:

$S_0$ - в СМО нет ни одной заявки;

$S_1$ - в СМО находится одна заявка (один канал занят, остальные свободны).

$\ldots$

$S_k$ - в СМО находится $k$ заявок ($k$ каналов заняты, остальные свободны).

$\ldots$

$S_n$ - в СМО находится $n$ заявок (все $n$ каналов заняты).

<center><img src="2.png" width=700 height=700></center>

Как видно из рисунка, интенсивность потоков обслуживания увеличивается с каждым занятым каналом. Используя формулы финальных вероятностей для схемы размножения гибели, получаем:
$$P_0=\left(1+\frac{\lambda}{\mu}+\frac{\lambda^2}{2!\mu^2}+\frac{\lambda^3}{3!\mu^3}+\ldots+\frac{\lambda^k}{k!\mu^k}+\ldots+\frac{\lambda^n}{n!\mu^n}\right)^{-1}$$

$$P_0=\left(\sum_{k=0}^n\frac{\lambda^k}{k!\cdot \mu^k}\right)^{-1}=\left(\sum_{k=0}^n\frac{y^k}{k!}\right)^{-1}\tag{2.1}$$

$$\forall k>0: P_k=\frac{y^k}{k!}\cdot P_0\tag{2.2}$$

Формулы (2.1), (2.2) и называются формулами Эрланга. По финальным вероятностям можно вычислить характеристики эффективности СМО.

**Вероятность того, что заявка получит отказ** совпадает с вероятностью, что все $n$ каналов будут заняты:
$$P_{\text{отк}}=P_n=\frac{y^n}{n!}\cdot P_0\tag{2.3}$$

**Вероятность обслуживания заявки** $$P_{\text{обсл}}=1-P_{\text{отк}}=1-\frac{y^n}{n!}\cdot P_0\tag{2.4}$$

**Абсолютная пропускная способность**:
$$\lambda'=\lambda\cdot \left(1-\frac{y^n}{n!}\cdot P_0\right)\tag{2.5}$$
Так как каждый занятый канал в единицу времени обслуживает в среднем $\mu$ заявок, **среднее число занятых каналов** равно:
$$k=\frac{\lambda'}{\mu}=y\cdot \left(1-\frac{y^n}{n!}\cdot P_0\right)\tag{2.6}$$

В качестве дополнительных показателей эффективности можно рассмотреть:
- **среднее число свободных каналов**:
$$k_{\text{св}}=n-k\tag{2.7}$$
- **коэффициент загрузки каналов** - отношение среднего числа занятых каналов к общему числу имеющихся каналов:
$$k_{\text{загр}}=\frac{k}{n}\tag{2.8}$$
- **коэффициент простоя каналов**:
$$k_{\text{пр}}=1-k_{\text{загр}}\tag{2.9}$$

**Пример 2.1.** В вычислительный центр коллективного пользования с тремя ЭВМ поступают заказы на вычислительные работы. Если работают все три ЭВМ, то вновь поступающий заказ не принимается, и предприятие вынуждено обратиться в другой вычислительный центр. Среднее время работы с одним заказом составляет 3 часа. Интенсивность потока заявок 0.25 заявок в час. Найти предельные вероятности и показатели эффективности работы вычислительного центра.

**Решение.**

По условию задачи $n=3, \lambda=0.25, \mu=\frac{1}{3}, y=\frac{\lambda}{\mu}=\frac{3}{4}$.  Найдем предельные вероятности:

$$P_0=\left(1+\frac{3}{4}+\frac{1}{2!}\cdot \left(\frac{3}{4}\right)^2+\frac{1}{3!}\cdot \left(\frac{3}{4}\right)^3\right)^{-1}$$


In [2]:
library(rim)
maxima.get("P0:(1+3/4+1/2!*(3/4)^2+1/3!*(3/4)^3)^(-1);")

(%o1) 128/269

In [5]:
maxima.get("float(P0);");

(%o3) 0.4758364312267658

В соответствии с формулой (2.3) рассчитываем остальные вероятности:


In [12]:
maxima.get("makelist(((3/4)^k/k!*P0),k,1,3);")

(%o8) [96/269,36/269,9/269]

In [13]:
maxima.get("float(%);")

(%o9) [0.35687732342007433,0.13382899628252787,0.03345724907063197]

Таким образом, в стационарном режиме работы вычислительного центра в среднем 47.6% времени нет ни одной заявки, 35.7% времени - имеется одна заявка, 13.4% времени - заняты две ЭВМ, 3.3% времени - заняты все три ЭВМ.

Следовательно, вероятность отказа равна $$P_{\text{отк}}=P_3\approx 0.033.$$

Абсолютная пропускная способность $\lambda'=0.25\cdot(1-0.033)\approx 0.24175$ заявки.

Среднее число занятых каналов $k=\frac{0.24175}{0.33}\approx 0.725$, т.е. каждая из трех ЭВМ будет занята обслуживанием заявок в среднем $\frac{72.5}{3}\approx 24.2\%$ времени.

При оценке эффективности работы вычислительного центра нужно сопоставить доходы от выполнения заявок с потерями от простоев дорогостоящих ЭВМ и принять компромиссное решение.

## $\S 3.$ **Одноканальная СМО с неограниченной очередью $M/M/1/\infty$**

### **Открытая система**

На практике довольно часто встречаются одноканальные СМО с очередью (ЭВМ, выполняющая запросы пользователя, врач, обслуживающий пациентов и т.д.). Если количество пользователей, от которых могут поступать запросы, никак не ограничено, то мы имеем дело с открытой системой.

Пусть имеется одноканальная СМО с неограниченной очередью (ни по длине очереди, ни по времени ожидания). На эту СМО поступает поток заявок с интенсивностью $\lambda$ и поток обслуживания имеет интенсивность $\mu$ ($\lambda, \mu$ сохраняют свои значения в течение всего времени функционирования системы). 

Требуется найти финальные вероятности состояний СМО, а также характеристики ее эффективности.

Состояния системы будем нумеровать по числу заявок, находящихся в СМО:
- $S_0$ - канал свободен;
- $S_1$ - канал занят, очереди нет;
- $S_2$ - канал занят, одна заявка стоит в очереди и т.д.

Число состояний ничем не ограничено. Тогда получаем классический процесс размножения, гибели при $\lambda=\lambda_k, \mu=\mu_k,$ но с бесконечным числом состояний. По всем стрелкам поток заявок с интенсивностью $\lambda$ переводит систему слева направо, а справа налево - поток обслуживания с интенсивностью $\mu$.

Если $\lambda\geq\mu$, то канал с заявками не справляется, очередь растет до бесконечности, если $\lambda<\mu$, то задача разрешима.

Обозначим $y=\frac{\lambda}{\mu}$, тогда, используя формулу (3.6) из лекции 3, получаем:
$$P_0=\left(1+y+y^2+y^3+\ldots+y^k+\ldots\right)^{-1}$$

Выражение $1+y+y^2+y^3+\ldots+y^k+\ldots=\frac{1}{1-y}, k\rightarrow \infty$ представляет собой геометрическую прогрессию, при $y<1$ ряд сходится, откуда получаем:
$$P_0=1-y\tag{3.1}$$

Остальные финальные вероятности находятся по формуле (3.5) лекции 3:
$$\forall k=1,\ldots,\infty: P_k=y^k\cdot P_0=y^k(1-y)\tag{3.2}$$

Из полученных формул, кстати, видно, что наибольшее значение имеет вероятность $P_0$ - того, что канал будет вообще свободен. Это связано с тем, что как бы ни была загружена система с очередью, если она вообще справляется с потоком заявок, самое вероятное число заявок в системе будет равно нулю.  

Найдем **среднее число заявок в СМО** $L_{\text{сист}}$. Случайная величина - число заявок в системе, имеет возможные значения $0, 1, 2, \ldots$ с вероятностями $P_0, P_1, P_2, \ldots$. Ее математическое ожидание:
$$L_{\text{сист}}=0\cdot P_0+1\cdot P_1+2\cdot P_2+\ldots = \sum_{k=1}^{\infty}k\cdot P_k=(1-y)\cdot y\cdot \sum_{k=1}^{\infty}k\cdot y^{k-1}=(1-y)\cdot y\cdot\frac{d}{d y}\sum_{k=1}^\infty y^k$$
$$L_{\text{сист}}=(1-y)\cdot y\cdot\frac{d}{d y}\left(\frac{y}{1-y}\right)=(1-y)\cdot y\cdot\frac{1}{(1-y)^2}$$
$$L_{\text{сист}}=\frac{y}{1-y}\tag{3.3}$$

По формуле Литтла найдем **среднее время пребывания заявки в системе**:
$$T_{\text{сист}}=\frac{L_{\text{сист}}}{\lambda}=\frac{y}{\lambda(1-y)}=\frac{1}{\mu(1-y)}=\frac{1}{\mu-\lambda}\tag{3.4}$$

**Среднее время ожидания обслуживания**:
$$W_{\text{оч}}=T_{\text{сист}}-\frac{1}{\mu}=\frac{1}{\mu(1-y)}-\frac{1}{\mu}=\frac{y}{\mu(1-y)}=\frac{y^2}{\lambda(1-y)}=\frac{\lambda}{\mu(\mu-\lambda)}\tag{3.5}$$

**Средняя длина очереди**:
$$L_{\text{оч}}=\lambda\cdot W_{\text{оч}}=\frac{y^2}{(1-y)}\tag{3.6}$$

### **Замкнутая система**

Если количество запросов, которые могут поступать в систему одномоментно как-то ограничено, то мы имеем дело с замкнутой системой (или системой Энгсета). 

Например, в серверной имеется $k$ серверов, требующих обслуживания (в среднем через $t_1$ мин. возникает ситуация, требующая обслуживания), имеется всего один сисадмин, который восстанавливает работоспособность серверов. Требуется оценить эффективность функционирования данной СМО.

В замкнутых системах у источников заявок принято различать два состояния: **активное**, когда источник может сгенерировать новую заявку на обслуживание, **пассивное**, когда источник послал в СМО заявку, но еще не получил от нее ответа. В пассивном состоянии источник не может сгенерировать новую задачу. Если количество источников заявок ограничено, то интенсивность входящего потока будет зависеть от того, сколько источников находятся в активном или пассивном состояниях. В то же время исследования показали, что эта зависимость потока заявок от состояния СМО проявляется только при небольшом числе источников заявок. Если число источников достаточно велико, то практически интенсивность общего потока заявок перестает зависеть от состояния СМО (например, количество абонентов у операторов мобильной связи хотя и ограничено, но настолько велико, что это практически не отражается на входящем потоке запросов от абонентов).

Обозначим интенсивность потока отказов одного сервера через $\lambda=\frac{1}{t_1}$, тогда $k$ серверов дают отказы с интенсивностью $k\cdot \lambda$. Пусть также сисадмин обслуживает серверы с интенсивностью $\mu$. Случайные величины: время между последовательными поломками серверов, время обслуживания сервера имеют экспоненциальное распределение.

Пусть $S_0$ - состояние исправности всех серверов; $S_1$ - сломался один сервер, очереди нет;
$S_2$ - сломалось два сервера, один в очереди и т.д.

<center><img src="ris1.png" width=400 height=400></center>

В этом случае по формуле (3.6) лекции 3 получаем:

$$P_0=\left(1+\frac{k\lambda}{\mu}+\frac{k(k-1)\lambda^2}{\mu^2}+\frac{k(k-1)(k-2)\lambda^3}{\mu^3}+\ldots+\frac{k(k-1)\cdot\ldots\cdot(k-m+1)\lambda^{m}}{\mu^{m}}+\ldots+\frac{k!\lambda^k}{\mu^k}\right)^{-1}$$

$$P_0=\left(\sum_{m=0}^{k}\frac{k!\lambda^m}{(k-m)!\mu^m}\right)^{-1}\tag{3.7}$$

$$P_1=P_0\cdot \frac{k\lambda}{\mu}$$
$$P_2=P_0\cdot \frac{k(k-1)\lambda^2}{\mu^2}$$
$$\ldots$$
$$P_m=P_0\cdot \frac{k(k-1)\ldots(k-m+1)\lambda^{m}}{\mu^{m}}$$
$$\ldots$$
$$P_k=P_0\cdot \frac{k!\lambda^k}{\mu^k}$$

Вероятность $P_m$ можно также записать в виде:
$$P_m=P_0\cdot \frac{k!\lambda^{m}}{(k-m)!\mu^{m}} \tag{3.8}$$

Заметим, что в данном случае система имеет конечное число состояний: $S_0, S_1, \ldots, S_k$

**Среднее число заявок в системе** равно среднему числу источников заявок, находящихся в пассивном состоянии:
$$L_{\text{сист}}=\sum_{m=0}^km\cdot P_m$$
$$L_{\text{сист}}=N_{\text{пас}}=k-N_{\text{акт}}=k-\frac{(1-P_0)\cdot \mu}{\lambda}\tag{3.9}$$
$(1-P_0)$ - вероятность, что канал занят;

$\frac{\mu}{\lambda}$ - отношение среднего количества обслуженных заявок к среднему количеству поступивших заявок.

**Средняя интенсивность суммарного входного потока**, порождаемого активными источниками, равна произведению среднего числа активных источников на интенсивность потока от каждого источника:
$$\Lambda=N_{\text{акт}}\cdot \lambda = (k-L_{\text{сист}})\cdot \lambda=(1-P_0)\cdot \mu\tag{3.10}$$

Поскольку все заявки будут обслужены, средняя интенсивность суммарного входного потока, т.е. среднее число поданных в единицу времени заявок, равна абсолютной пропускной способности системы:
$$\lambda'=\Lambda$$

**Среднее время пребывания заявки в системе**:

В случае замкнутых систем поток заявок имеет меняющуюся в зависимости от состояния системы интенсивность потока. Поэтому в таких системах вместо интенсивности отдельного источника заявок $\lambda$ необходимо взять
среднюю интенсивность суммарного потока заявок $\Lambda$.
$$L_{\text{сист}}=\Lambda\cdot T_{\text{сист}}\Rightarrow \\ T_{\text{сист}}=\frac{L_{\text{сист}}}{\Lambda}\tag{3.11}$$

**Среднее число занятых каналов** равно среднему числу заявок, находящихся под обслуживанием (находим как математическое ожидание):
$$y=0\cdot P_0+1\cdot (1-P_0)=1-P_0 \tag{3.12}$$

**Среднее длина очереди** найдем как разность среднего числа заявок, находящихся в системе, и среднего числа заявок, находящихся под обслуживанием:
$$L_{\text{оч}}=L_{\text{сист}}-y=k-\frac{(1-P_0)\mu}{\lambda}-(1-P_0)=k-\frac{(1-P_0)(\lambda+\mu)}{\lambda}\tag{3.13}$$

**Среднее время нахождения в очереди**:
$$W_{\text{оч}}=\frac{L_{\text{оч}}}{\Lambda}\tag{3.14}$$

## $\S 4.$ **Многоканальная СМО с неограниченной очередью $M/M/n/\infty$**

Аналогично одноканальной СМО решается задача и с многоканальной СМО с неограниченной очередью и $n$ каналами. Введем следующие состояния:

$S_0$ - все каналы свободны;

$S_1$ - один канал занят, очереди нет;

$S_2$ - два канала заняты, очереди нет;

$\ldots$

$S_n$ - занято $n$ каналов, очереди нет;

$S_{n+1}$ - занято $n$ каналов, одна заявка стоит в очереди;

$\ldots$

$S_{n+r}$  - занято $n$ каналов, $r$ заявок стоит в очереди;

$\ldots$

<center><img src="ris2.png" width=700 height=700></center>

Полученный граф состояний соответствует схеме размножения гибели с бесконечным количеством состояний. Естественное условие существования финальных вероятностей  - это $\frac{\lambda}{n\mu}<1$.

Применяя формулы размножения гибели, найдем финальные вероятности.

Обозначим $y=\frac{\lambda}{\mu}$, получим:
 
$$P_0=\left(1+\frac{y}{1!}+\frac{y^2}{2!}+\frac{y^3}{3!}+\ldots+\frac{y^n}{n!}+\frac{y^n\cdot y}{n!n}+\frac{y^n\cdot y^2}{n!n^2}+\frac{y^n\cdot y^3}{n!n^3}+\ldots\right)^{-1}$$

$$P_0=\left(\sum_{i=0}^n\frac{y^i}{i!}+\frac{y^{n+1}}{n!n}\left(1+\frac{y}{n}+\frac{y^2}{n^2}+\ldots\right)\right)^{-1}=\left(\sum_{i=0}^n\frac{y^i}{i!}+\frac{y^{n+1}}{n!n}\frac{n}{n-y}\right)^{-1}$$

$$P_0=\left(\sum_{i=0}^n\frac{y^i}{i!}+\frac{y^{n+1}}{n!(n-y)}\right)^{-1}\tag{4.1}$$

из формулы (3.5) лекции 3 получаем:
$$P_1=\frac{y}{1!}\cdot P_0, P_2=\frac{y^2}{2!}\cdot P_0, \ldots, P_k=\frac{y^k}{k!}\cdot P_0, \ldots, P_n=\frac{y^n}{n!}\cdot P_0\tag{4.2}$$
$$P_{n+1}=\frac{y^{n+1}}{n\cdot n!}\cdot P_0, \ldots, P_{n+r}=\frac{y^{n+r}}{n^r\cdot n!}\cdot P_0\tag{4.3}$$

Найдем основные характеристики эффективности СМО. 

**Абсолютная пропускная способность** равна интенсивности входящего потока заявок:

$$\lambda'=\lambda$$

**Среднее число заявок в очереди**:

очередь начинает образовываться, когда занятыми оказываются все $n$ каналов:
$$L_{\text{оч}}=\sum_{r=1}^\infty r\cdot P_{n+r}=\sum_{r=1}^\infty r\cdot \frac{y^{n+r}}{n^r\cdot n!}P_0$$

$$L_{\text{оч}}=\frac{y^{n+1}}{n!\cdot n}\cdot P_0\sum_{r=1}^{\infty}\frac{r\cdot y^{r-1}}{n^{r-1}}=\frac{y^{n+1}}{n!\cdot n}\cdot P_0\cdot \frac{d}{d(y/n)}\sum_{r=1}^{\infty}\frac{y^r}{n^r}=\frac{y^{n+1}}{n!\cdot n}\cdot P_0\cdot \frac{d}{d(y/n)}\left(\frac{y/n}{1-y/n}\right)$$
$$L_{\text{оч}}=\frac{y^{n+1}}{n!\cdot n}\cdot P_0\cdot \frac{1}{(1-y/n)^2}=\frac{y^{n+1}P_0}{(n-1)!(n-y)^2}\tag{4.4}$$

**Среднее число заявок в системе** равно сумме среднего числа заявок в очереди и среднего числа заявок под обслуживанием (оно же среднее число занятых каналов). В свою очередь, среднее число занятых каналов при условии $\frac{\lambda}{n\mu}<1$ равно отношению интенсивности входящего потока $\lambda$ к интенсивности обслуживания одного канала $\mu$: $y=\frac{\lambda}{\mu}$:
$$L_{\text{сист}}=L_{\text{оч}}+y\tag{4.5}$$

По формулам Литтла легко находим:
- **среднее время нахождения заявки в очереди**:
$$W_{\text{оч}}=\frac{L_{\text{оч}}}{\lambda}\tag{4.6}$$
- **среднее время нахождения заявки в системе**:
$$T_{\text{сист}}=\frac{L_{\text{сист}}}{\lambda}\tag{4.7}$$

## $\S 5.$ **Одноканальная СМО с ограничением на длину очереди $M/M/1/(m+1)$**

Ограничение на длину очереди означает, что в очереди имеется всего $m$ мест, и заявка, поступившая, когда все $m$ мест заняты, отклоняется и покидает систему необслуженной. Граф состояний для одноканальной системы с ограничением на длину очереди представлен на рисунке.

<center><img src="ris5.png" width=700 height=700></center>


Как видно из рисунка, возможны следующие состояния:

$S_0$ - канал обслуживания свободен;

$S_1$ - канал обслуживания занят, очереди нет;

$S_2$ - канал обслуживания занят, в очереди одна заявка;

$\ldots$

$S_{k+1}$ - канал обслуживания занят, в очереди $k$ заявок;

$S_{m+1}$ - канал обслуживания занят, в очереди $m$ заявок.

**Вероятность отказа** в такой системе равна:

$$P_{\text{отк}}=P_{m+1}=y^{m+1}P_0\tag{5.1}$$

**Абсолютная пропускная способность** равна:
$$\lambda'=(1-P_{\text{отк}})\lambda=\lambda-\lambda y^{m+1}P_0\tag{5.2}$$

Остальные параметры находятся в зависимости от значения $y$.

Если $y=1$, то формулы получаются простыми:
$$P_0=\left(1+\sum_{k=1}^{m+1}y^k\right)^{-1}=\frac{1}{m+2}\tag{5.3}$$
$$\forall k>0: P_k=y^{k}P_0=\frac{1}{m+2}\tag{5.4}$$

**Средняя длина очереди**:

$$L_{\text{оч}}=\sum_{k=1}^{m}kP_{k+1}=\frac{m(m+1)}{2}P_0=\frac{m(m+1)}{2(m+2)}\tag{5.5}$$

**Среднее время ожидания в очереди**:
$$W_{\text{оч}}=\frac{L_{\text{оч}}}{\lambda'}\tag{5.6}$$
В системе с ограниченной очередью ($m$ мест) заявки могут получать отказ, если очередь заполнена. Поэтому не все заявки, приходящие в систему (с интенсивностью $\lambda$), реально в нее попадают, поэтому делим на пропускную способность.

**Среднее число заявок, находящихся под обслуживанием**:
$$L_{\text{обс}}=0\cdot P_0+1\cdot (1-P_0)=\frac{m+1}{m+2}\tag{5.7}$$

**Среднее число заявок в системе**:
$$L_{\text{сист}}=L_{\text{обс}}+L_{\text{оч}}=\frac{m+1}{m+2}+\frac{m(m+1)}{2(m+2)}=\frac{m+1}{2}\tag{5.8}$$
**Среднее время пребывания заявки в системе** включает в себя время обслуживания и время ожидания в очереди:
$$T_{\text{сист}}=\frac{L_{\text{сист}}}{\lambda'}\tag{5.9}$$

Если $y\neq 1$, то

**Вероятность простоя системы**:
$$P_0=\left(\sum_{m=0}^{m+1}y^m\right)^{-1}=\left(\frac{y^{m+2}-1}{y-1}\right)^{-1}=\frac{y-1}{y^{m+2}-1}\tag{5.10}$$
[Wolfram](https://www.wolframalpha.com/input?i=sum%28y%5Ek%2Ck%2C0%2Cm%2B1%29)
$$\forall k>0:P_k=y^kP_0\tag{5.11}$$

**Среднее число заявок в очереди**
$$L_{\text{оч}}=\sum_{k=1}^{m}kP_{k+1}=\frac{y-1}{y^{m+2}-1}\cdot\sum_{k=1}^m ky^{k+1}=\frac{y^2(y-1)}{y^{m+2}-1}\frac{d}{dy}\sum_{k=1}^m y^k=\frac{y^2(y-1)}{y^{m+2}-1}\frac{d}{dy}\left(\frac{y(y^m-1)}{y-1}\right)$$

[Wolfram - сумма геометрической прогрессии](https://www.wolframalpha.com/input?i=sum%28y%5Ek%2Ck%2C1%2Cm%29)

[Wolfram - производная](https://www.wolframalpha.com/input?i=diff%28sum%28y%5Ek%2Ck%2C1%2Cm%29%2Cy%29)

$$L_{\text{оч}}=\frac{y^2(y-1)}{y^{m+2}-1}\cdot\frac{my^{m+1}-(m+1)y^m+1}{(y-1)^2}$$

$$L_{\text{оч}}=\frac{y^2}{y^{m+2}-1}\cdot\frac{my^{m+1}-(m+1)y^m+1}{(y-1)}\tag{5.12}$$

**Среднее время ожидания в очереди**:
$$W_{\text{оч}}=\frac{L_{\text{оч}}}{\lambda'}\tag{5.13}$$
**Среднее число заявок в системе**:

среднее число заявок, находящихся под обслуживанием (средняя загрузка канала):
$$L_{\text{обс}}=1-P_0$$
$$L_{\text{сист}}=L_{\text{оч}}+L_{\text{обс}}\tag{5.14}$$
**Среднее время пребывания заявки в системе** включает в себя время обслуживания и время ожидания в очереди:
$$T_{\text{сист}}=\frac{L_{\text{сист}}}{\lambda'}\tag{5.15}$$


## $\S 6.$ **Многоканальная СМО с ограничением на длину очереди $M/M/n/(m+n)$**

Пусть на вход системы массового обслуживания, имеющей $n$ каналов, поступает пуассоновский поток с интенсивностью $\lambda$. Интенсивность обслуживания каждой заявки равна  $\mu$, а максимальное число мест в очереди равно $m$.

<center><img src="ris6.png" width=700 height=700></center>

$S_0$ - система свободна;

$S_1$ - занят один канал, очереди нет;

$S_2$ - занято два канала, очереди нет;

$\ldots$

$S_n$ - заняты все каналы, очереди нет;

$S_{n+1}$ - заняты все каналы, одна заявка в очереди

$\ldots$

$S_{n+m}$ - заняты все каналы, заняты все $m$ мест очереди.


**Вероятность простоя**:
$$P_0=\left(1+\frac{y}{1!}+\frac{y^2}{2!}+\ldots+\frac{y^n}{n!}+\frac{y^n\cdot y}{n!n}+\frac{y^n\cdot y^2}{n!n^2}+\frac{y^n\cdot y^3}{n!n^3}+\ldots+\frac{y^n\cdot y^m}{n!n^m}\right)^{-1}$$
$$P_0=\left(\sum_{k=0}^n \frac{y^k}{k!}+\frac{y^n}{n!}\left(\frac{y}{n}+\frac{y^2}{n^2}+\frac{y^3}{n^3}+\ldots +\frac{y^m}{n^m}\right)\right)^{-1}$$

[Wolfram - упрощения](https://www.wolframalpha.com/input?i=sum%28y%5Ek%2Fn%5Ek%2Ck%2C1%2Cm%29)


$$P_0=\left(\sum_{k=0}^n \frac{y^k}{k!}+\frac{y^{n+1}}{n!\cdot n}\cdot \frac{1-\left(\frac{y}{n}\right)^{m}}{1-\frac{y}{n}}\right)^{-1}\tag{6.1}$$
$$P_{n+1}=\frac{y^{n+1}}{n\cdot n!}\cdot P_0, \ldots, P_{n+m}=\frac{y^{n+m}}{n^m\cdot n!}\cdot P_0\tag{6.2}$$

$$P_{i\leq n}=\frac{y^{i}}{i!}\cdot P_0 : P_{1}=\frac{y^{1}}{1!}\cdot P_0, \ldots, P_{n}=\frac{y^{n}}{n!}\cdot P_0\tag{6.3}$$

**Вероятность образования очереди**:
$$P_{n+1}+P_{n+2}+\ldots+P_{n+m}=\frac{y^n}{n!}P_0\left(\frac{y}{n}+\frac{y^2}{n^2}+\ldots+\frac{y^m}{n^m}\right)=\frac{y^{n+1}}{n!\cdot n}\cdot P_0\cdot \frac{1-\left(\frac{y}{n}\right)^m}{1-\frac{y}{n}}\tag{6.4}$$
**Вероятность отказа**:
$$P_{\text{отк}}=P_{n+m}=\frac{y^{n+m}}{n^m\cdot n!}\cdot P_0\tag{6.5}$$

Найдем основные характеристики эффективности СМО. 

**Абсолютная пропускная способность**:

$$\lambda'=(1-P_{\text{отк}})\lambda\tag{6.6}$$

**Среднее число заявок в очереди**:

очередь начинает образовываться, когда занятыми оказываются все $n$ каналов:
$$L_{\text{оч}}=\sum_{k=1}^m k\cdot P_{n+k}=\sum_{k=1}^m k\frac{y^{n+k}}{n^k\cdot n!}\cdot P_0=\frac{y^nP_0}{n!}\sum_{k=1}^m \frac{ky^k}{n^k}$$
[Wolfram - упрощение суммы](https://www.wolframalpha.com/input?i=sum%28k*y%5Ek%2Fn%5Ek%2Ck%2C1%2Cm%29)
$$L_{\text{оч}}=\frac{y^nP_0}{n!}\cdot \frac{yn^{-m}(n^{m+1}-(m+1)ny^m+my^{m+1})}{(n-y)^2}\tag{6.7}$$

**Среднее число занятых каналов**:
$$k=\frac{\lambda'}{\mu}=y\cdot(1-P_{\text{отк}})\tag{6.8}$$

**Среднее число заявок в системе** равно сумме среднего числа заявок в очереди и среднего числа заявок под обслуживанием (оно же среднее число занятых каналов):
$$L_{\text{сист}}=L_{\text{оч}}+y(1-P_{\text{отк}})\tag{6.9}$$

По формулам Литтла легко находим:
- **среднее время нахождения заявки в очереди**:
$$W_{\text{оч}}=\frac{L_{\text{оч}}}{\lambda'}\tag{6.10}$$
В системе с ограниченной очередью ($m$ мест) заявки могут получать отказ, если очередь заполнена. Поэтому не все заявки, приходящие в систему (с интенсивностью $\lambda$), реально в нее попадают, поэтому делим на пропускную способность.
- **среднее время нахождения заявки в системе**:
$$T_{\text{сист}}=\frac{L_{\text{сист}}}{\lambda'}\tag{6.11}$$

## $\S 7.$ **Описание пакета моделирования СМО queueing в системе R**

Пакет queueing в R предназначен для аналитического решения и расчета основных показателей производительности систем массового обслуживания и сетей массового обслуживания . Он предоставляет инструменты для работы с марковскими моделями, где входящий поток и обслуживание являются пуассоновскими (экспоненциальное распределение интервалов). Он также позволяет работать с одно- и многоклассовыми сетями (будем рассматривать в дальнейшем). Основное преимущество пакета — аналитический подход, который обеспечивает быстрое и точное вычисление характеристик без затрат на моделирование.

Процесс работы с пакетом состоит из нескольких стандартных шагов:

1. Создание входных данных: используется функция **NewInput.<модель>**.

2. Проверка данных (опционально): вызывается функция **CheckInput()**.

3. Решение модели: вызывается функция **QueueingModel()**.

4. Получение результатов: вызывается функция **summary** от модели.

Параметры, которые задаются функцией NewInput:

- **lambda** ($\lambda$): Интенсивность входящего потока - среднее число заявок (клиентов, вызовов), поступающих в систему в единицу времени.

- **mu** ($\mu$): Интенсивность обслуживания - среднее число заявок, которое один сервер (канал, оператор) способен обслужить в единицу времени.

- **c**: Число серверов (каналов) - Количество параллельно работающих обслуживающих приборов (кассиров, процессоров, ремонтников).

- **k**: Емкость системы - максимальное количество заявок, которое может одновременно находиться в системе (те, что на обслуживании + те, что в очереди). Если это число достигнуто, новые заявки получают отказ. В моделях без ограничений (например, M/M/c) этот параметр равен бесконечности и не указывается.

- **m**: Размер генеральной совокупности (источник нагрузки) - общее количество потенциальных клиентов, которые могут когда-либо прийти в систему. Например, количество станков, которые могут сломаться, или количество сотрудников в офисе. Используется в замкнутых моделях (например, M/M/c/k/m).

Для анализа результатов: используются функции summary(), print(), а также специализированные функции для получения отдельных показателей:

- **RO (или rho, ρ)**: Коэффициент использования системы (загрузка) - вероятность того, что случайно выбранный сервер занят, или доля времени, которую сервер работает. Показывает, насколько загружена система. Для одноканальной системы: $\rho = \frac{\lambda}{\mu}$. Для многоканальной: $\rho = \frac{\lambda}{c\cdot\mu}$. Система работает устойчиво (не уходит в бесконечную очередь), если $\rho$.

- **P0**: Вероятность простоя системы - вероятность того, что в системе нет ни одной заявки (все серверы свободны, очередь пуста).

- **L (Mean number in system)**: Среднее число заявок в системе - сколько в среднем клиентов находится в системе в данный момент времени (и в очереди, и на обслуживании).

- **Lq (Mean number in queue)**: Средняя длина очереди - сколько в среднем клиентов стоит в очереди (ожидает обслуживания).

- **W (Mean waiting time in system)**: Среднее время пребывания в системе - сколько времени в среднем проходит с момента прихода клиента до момента его полного ухода из системы (ожидание + обслуживание).

- **Wq (Mean waiting time in queue)**: Среднее время ожидания в очереди - сколько времени в среднем клиент проводит непосредственно в очереди, прежде чем попадет на обслуживание.

- **X (Throughput)**: Пропускная способность (интенсивность выходящего потока) - среднее число заявок, покидающих систему (обслуженных) в единицу времени. В установившемся режиме для разомкнутых систем (открытых моделей) X обычно равен lambda (сколько вошло, столько и вышло, если нет потерь). Для замкнутых систем это ключевой показатель производительности.

- **Wqq**: понимается среднее время ожидания в очереди для заявок, которые действительно стоят в очереди (conditional waiting time). Простой Wq считает среднее время ожидания по всем заявкам, включая те, которые вообще не ждали (обслужились сразу). А Wqq считает среднее время ожидания только для тех заявок, которые вынуждены были ждать.

- **Lqq**: это среднее число клиентов в очереди при условии, что очередь существует (то есть не пуста). Простыми словами, это ответ на вопрос: «Когда в очереди уже есть люди, сколько их обычно там стоит?». Главное отличие от стандартного показателя Lq (средняя длина очереди) в том, что Lq усредняет состояние очереди по всему времени, включая периоды, когда очередь пуста. Lqq же смотрит только на те моменты, когда очередь не пуста, и вычисляет среднее для этих моментов.

- **VNq (Variance of Number in Queue)**: это дисперсия (разброс) длины очереди. Насколько сильно фактическое число людей в очереди колеблется относительно среднего значения Lq. Высокое значение означает, что очередь то пуста, то внезапно вырастает до огромных размеров.

- **VTq (Variance of Time in Queue)**: это дисперсия (разброс) времени ожидания в очереди. Насколько сильно отличается время ожидания разных клиентов. Если значение большое, это значит, что кому-то повезло (обслужился сразу), а кто-то простоял очень долго.

- **VN (Variance of Number in System)**: это дисперсия (разброс) числа заявок в системе (в очереди и на обслуживании).

- **VT (Variance of Time in System)**: это дисперсия (разброс) времени пребывания в системе (ожидание + обслуживание).

- **Pn (Probability of being in state n)**: это стационарная вероятность того, что в системе находится ровно n заявок (клиентов). Если вы посмотрите на систему в случайный момент времени, то Pn — это вероятность застать в ней n человек. Например, P0 — это вероятность того, что система пуста (частный случай).

- **Qn (Cumulative Probability of n or more)**: это вероятность того, что в системе находится n или более заявок, т.е. это "хвост" распределения. Рассчитывается обычно как 1 - (P0 + P1 + ... + Pn-1). Например, если вы хотите знать вероятность того, что очередь будет больше 10 человек, это и есть Q11.

Также пакет может возвращать сами функции распределения. Это, пожалуй, самая мощная аналитическая возможность пакета. В отличие от имитационного моделирования, где нужно строить гистограммы, здесь пакет может выдать точную формулу распределения вероятностей (CDF — Cumulative Distribution Function).

- **FW (Function of Waiting time in System)**: это функция распределения времени пребывания в системе P(W <= t). В R вы можете вызвать её и передать ей аргумент t (время), чтобы получить вероятность того, что клиент проведет в системе не больше t времени. Пример: FW(o_model, t = 5) вернет вероятность того, что время пребывания в системе (ожидание + обслуживание) не превысит 5 единиц времени.

- **FWq (Function of Waiting time in Queue)**: это функция распределения времени ожидания в очереди P(Wq <= t). FWq(o_model, t = 2) вернет вероятность того, что клиент простоит в очереди не больше 2 единиц времени.


### Система класса $M/M/1/\infty$

In [3]:
# Загрузка пакета
library(queueing)

# 1. Определение входных параметров для M/M/1
# lambda = 2 (интенсивность поступления), mu = 3 (интенсивность обслуживания)
i_mm1 <- NewInput.MM1(lambda = 2, mu = 3)

# 2. Проверка введенных данных (необязательно, но рекомендуется)
CheckInput(i_mm1)

# 3. Создание и решение модели
o_mm1 <- QueueingModel(i_mm1)

# 4. Вывод сводки по основным показателям производительности
print(summary(o_mm1), digits = 2)

  lambda mu c  k  m   RO   P0  Lq   Wq X L W Wqq Lqq
1      2  3 1 NA NA 0.67 0.33 1.3 0.67 2 2 1   1   3


In [10]:
#Вероятность, что время пребывания в системе не превысит 3 единиц времени
o_mm1$FW(t=3)

#Вероятность, что время ожидания в очереди не превысит 2 единиц времени
o_mm1$FW(t=2)

[1] 0.9502129

[1] 0.8646647

Выполним расчеты согласно полученным ранее формулам:

$$P_0=1-y$$
$$W_{\text{оч}}=\frac{\lambda}{\mu(\mu-\lambda)}$$
$$L_{\text{оч}}=\lambda\cdot W_{\text{оч}}=\frac{y^2}{(1-y)}$$
$$L_{\text{сист}}=\frac{y}{1-y}$$
$$T_{\text{сист}}=\frac{1}{\mu(1-y)}=\frac{1}{\mu-\lambda}$$

In [7]:
lambda <- 2; mu <- 3
R0<-lambda/mu
cat('R0:',as.character(R0),'\n')
P0 <-1-R0
cat('P0:',as.character(P0),'\n')
Lq<-R0^2/(1-R0)
cat('Lq:',as.character(Lq),'\n')
Wq<-lambda/(mu*(mu-lambda))
cat('Wq:',as.character(Wq),'\n')
L<-R0/(1-R0)
cat('L:',as.character(L),'\n')
W <-1/(mu-lambda)
cat('W:',as.character(W),'\n')

R0: 0.666666666666667 
P0: 0.333333333333333 
Lq: 1.33333333333333 
Wq: 0.666666666666667 
L: 2 
W: 1 


Чтобы посмотреть возможные модели, можно использовать:

In [12]:
help.search("NewInput")

R Information

Help files with alias or concept or title matching ‘NewInput’ using
fuzzy matching:


queueing::NewInput.BnD
                        Define the inputs of a new generic Birth and
                        Death process model
  Aliases: NewInput.BnD
queueing::NewInput.CJN
                        Define the inputs of a Closed Jackson Network
  Aliases: NewInput.CJN, NewInput2.CJN, NewInput3.CJN
queueing::NewInput.MCCN
                        Define the inputs of a MultiClass Closed
                        Network
  Aliases: NewInput.MCCN
queueing::NewInput.MCMN
                        Define the inputs of a MultiClass Mixed Network
  Aliases: NewInput.MCMN
queueing::NewInput.MCON
                        Define the inputs of a MultiClass Open Network
  Aliases: NewInput.MCON
queueing::NewInput.MM1
                        Define the inputs of a new M/M/1 queueing model
  Aliases: NewInput.MM1
queueing::NewInput.MM1K
                        Define the inputs of a new M/M/1/K que

Чтобы получить справку по конкретной модели, можно использовать:

In [15]:
help(NewInput.MMCC)

NewInput.MMCC {queueing},R Documentation
lambda,arrival rate
mu,server service rate
c,number of servers
method,"with method = 0, the state probabilities are calculated using the formal definition (with overflow problems with factorials; with method = 1 (default), the truncated poisson distribution is used (recomended for professional use)"


В целом поддерживаются следующие модели:

$M/M/1$: Одноканальная система с бесконечной очередью.

$M/M/c$: Многоканальная система с бесконечной очередью.

$M/M/\infty$: Система с неограниченным числом приборов (без ожидания).

$M/M/1/K$: Одноканальная система с буфером ограниченной ёмкости (K мест).

$M/M/c/K$: Многоканальная система с буфером ограниченной ёмкости.

$M/M/c/c$: Многоканальная система с потерями (нет очереди).

$M/M/1/K/K$: Одноканальная система с конечным источником нагрузки (K источников).

$M/M/c/K/K$: Многоканальная система с конечным источником нагрузки.

$M/M/c/K/m$: Многоканальная система с конечным источником и ограниченным накопителем.

$M/M/\infty/K/K$: Система с неограниченным числом приборов и конечным источником.

### Система класса $M/M/c/c$

Данная система соответствует многоканальной СМО с отказами - в ней $c$ каналов и ровно $c$ заявок может находиться в системе. Именно такая задача и была решена Эрлангом. Рассмотрим пример:

**Пример 7.1.** Мы проектируем небольшую  линию поддержки клиентов компании. Нам известно, что в час пик поступает 180 звонков. Средняя длительность разговора составляет 2 минуты. Мы хотим, чтобы вероятность того, что звонок получит отказ (сигнал "Занято") была не более 2% (0.02).

Вопрос: Сколько линий (каналов/операторов) нам нужно установить?

**Решение.**

Максимальная интенсивность входного потока:

In [1]:
lambda <- 180/60
lambda

[1] 3

$\lambda = 3 \text{звонка в минуту}$

Интенсивность обслуживания:

In [2]:
mu <- 1/2
mu

[1] 0.5

In [18]:
i_mmcc <- NewInput.MMCC(lambda=lambda, mu=mu, c=5)

o_mmcc <- QueueingModel(i_mmcc)

In [19]:
str(o_mmcc)

List of 17
 $ Inputs    :List of 4
  ..$ lambda: num 3
  ..$ mu    : num 0.5
  ..$ c     : num 5
  ..$ method: num 1
  ..- attr(*, "class")= chr "i_MMCC"
 $ RO        : num 0.768
 $ Lq        : num 0
 $ VNq       : num 0
 $ Wq        : num 0
 $ VTq       : num 0
 $ Throughput: num 1.92
 $ L         : num 3.84
 $ VN        : num 1.32
 $ W         : num 2
 $ VT        : num 4
 $ Lqq       : logi NA
 $ Wqq       : logi NA
 $ Pn        : num [1:6] 0.00556 0.03337 0.10011 0.20022 0.30033 ...
 $ Qn        : num [1:5] 0.0087 0.0522 0.1565 0.313 0.4696
 $ FW        :function (t)  
 $ FWq       :function (t)  
 - attr(*, "class")= chr "o_MMCC"


Вероятность отказа - это вероятность, что все каналы будут заняты, в нашем случае это соответствует $Pn[c+1]$

In [22]:
o_mmcc$Pn[5]

[1] 0.3003337

Таким образом, при пяти каналах данная вероятность $0.3$. Значит нам нужно больше каналов.

In [32]:
for (c in seq(5,15)){
i_mmcc <- NewInput.MMCC(lambda=lambda, mu=mu, c=c)
o_mmcc <- QueueingModel(i_mmcc)
cat(as.character(o_mmcc$Pn[c+1]),' : ',as.character(c),'\n')}

0.360400444938821  :  5 
0.264922322158626  :  6 
0.185054735840076  :  7 
0.121875783666304  :  8 
0.0751449555469656  :  9 
0.0431418384104393  :  10 
0.0229908922111388  :  11 
0.0113648026294422  :  12 
0.00521792397844236  :  13 
0.00223126346370552  :  14 
0.000891709529924463  :  15 


Получается, что необходимо не менее 12 операторов.

Попробуем рассчитать аналитически:

$$P_0=\left(\sum_{k=0}^n\frac{\lambda^k}{k!\cdot \mu^k}\right)^{-1}=\left(\sum_{k=0}^n\frac{y^k}{k!}\right)^{-1}$$
$$\forall k>0: P_k=\frac{y^k}{k!}\cdot P_0$$

In [11]:
y <- lambda/mu
for (n in seq(5,15)){
     P0<-sum(sapply(1:n, function (k) y^k/factorial(k)))^(-1);
     Pn<-y^n/factorial(n)*P0;
     cat(as.character(Pn),' : ',as.character(n),'\n')}

0.36241610738255  :  5 
0.266009852216749  :  6 
0.18567335243553  :  7 
0.122233400402414  :  8 
0.0753488372093023  :  9 
0.0432538269846921  :  10 
0.0230491969141267  :  11 
0.0113932953035868  :  12 
0.00523093739498361  :  13 
0.00223681575078306  :  14 
0.000893926480780325  :  15 


## Система класса $M/M/c$

**Пример 7.2.**

В компании открывается служба поддержки интернет-провайдера. Анализ показал, что в час пик (с 19:00 до 20:00) поступает 240 звонков. Средняя длительность разговора с оператором составляет 3 минуты. Звонки не сбрасываются при занятости (клиенты готовы ждать на линии).

Руководство требует, чтобы уровень обслуживания (Service Level) составлял 80/20:

80% звонков должны быть приняты оператором в течение 20 секунд (т.е. время ожидания в очереди не более 20 секунд).

Сколько операторов необходимо нанять, чтобы обеспечить требуемый уровень сервиса в час пик?

**Решение.**

По условию задачи имеем систему с фиксированным количеством каналов и бесконечной очередью.

In [12]:
lambda <- 240/60
lambda

[1] 4

In [14]:
mu <- 1/3
mu

[1] 0.3333333

In [40]:
# Время ожидания в очереди для 80% звонков
20/60

[1] 0.3333333

In [37]:
c<-14

In [38]:
library(queueing)
i_mmc <- NewInput.MMC(lambda=lambda, mu=mu, c=c)
o_mmc <- QueueingModel(i_mmc)

Нас интересует время именно заявок, стоящих в очереди:

In [39]:
o_mmc$Wqq

[1] 1.5

Видно, что в среднем для всех среднее время при 14 каналах больше 0.333. Значит количество каналов нужно увеличивать. Для того, чтобы окончательно ответить на вопрос задачи, целесообразно рассмотреть функцию распределения FWq - функция распределения времени ожидания в очереди. Тогда FWq(o_mmc, t = 0.33) вернет вероятность того, что клиент простоит в очереди не больше 0.33 единиц времени. И нам нужно, чтобы эта вероятность была не меньше 0.8.

In [32]:
o_mmc$FWq(t=0.333)

[1] 0.9900692

In [36]:
for (c in seq(14,30)){
    i_mmc <- NewInput.MMC(lambda=lambda, mu=mu, c=c)
    o_mmc <- QueueingModel(i_mmc)
    cat(as.character(o_mmc$FWq(t=0.333)),' : ',as.character(c),'\n')}

0.614194765990104  :  14 
0.77121381701943  :  15 
0.86877294380825  :  16 
0.92730517435819  :  17 
0.961152190181094  :  18 
0.979989375161762  :  19 
0.990069195367659  :  20 
0.995252381580806  :  21 
0.997813277632006  :  22 
0.999029276024122  :  23 
0.999584456904843  :  24 
0.999828350090523  :  25 
0.999931529379214  :  26 
0.99997360348574  :  27 
0.999990156935415  :  28 
0.999996446808948  :  29 
0.999998757284341  :  30 


Таким образом, нам подходит 16 операторов.

## Система класса $M/M/c/c/N$

**Пример 7.3.**

В небольшом офисе работает 20 сотрудников. Каждому сотруднику иногда требуется переговорная комната для проведения совещаний. В час пик (обычно утром) каждый сотрудник в среднем пытается забронировать переговорку 1 раз в 2 часа.
Средняя длительность совещания составляет 30 минут. Если все переговорки заняты, сотрудник получает отказ (в этот раз совещание не проводится) и уходит работать дальше (возвращается в "пул" источников).

Необходимо, чтобы вероятность отказа (того, что сотрудник не сможет забронировать комнату) не превышала 5%. Сколько переговорных комнат (c) нужно оборудовать в офисе?

**Решение.**

Это многоканальная замкнутая СМО с отказами: $N=20$

In [44]:
#Интенсивность входящего потока
lambda <- 1/(2*60)

#Интенсивность обслуживания
mu <- 1/30

#Количество сотрудников
N<- 20

In [41]:
help(NewInput.MMCKM)

NewInput.MMCKM {queueing},R Documentation
lambda,arrival rate
mu,server service rate
c,number of servers
k,system capacity
m,"poblation size. Please, observe that should be m >= k"
method,"method of computation of the probabilities of k (system capacity) customers down. With method=0, the exact results are calculated using the formal definition. With method=1, aproximate results are calculated using Stirling aproximation of factorials and logaritms."


In [46]:
for (c in seq(2,10)){
    i_mmckm <- NewInput.MMCKM(lambda=lambda, mu=mu, c=c, k=c,m=N)
    o_mmckm <- QueueingModel(i_mmckm)
    cat(as.character(o_mmckm$Pn[c+1]),' : ',as.character(c),'\n')}

0.664335664335664  :  2 
0.499124343257443  :  3 
0.346541735212074  :  4 
0.217057736461892  :  5 
0.119455607879879  :  6 
0.0563614578365111  :  7 
0.0223843121812808  :  8 
0.00740617667020837  :  9 
0.00203255887452503  :  10 


Таким образом, для достижения цели достаточно 8 переговорок.

**Пример 7.3.**

Имеется серверный кластер для обработки вычислительных задач. Задачи поступают в случайные моменты времени и при занятости всех серверов становятся в очередь ограниченной длины. Если очередь заполнена, задача получает отказ.

Параметры системы:

- Интенсивность поступления задач: λ = 100 задач/мин
- Производительность одного сервера: μ = 40 задач/мин
- В кластере c = 2 сервера
- Размер буфера (очереди): m = 10 мест
- Каждая выполненная задача приносит доход 5 руб.
- Содержание одного сервера обходится 2000 руб/час
- Время работы кластера: 24 часа/сутки

Технические требования:

Вероятность отказа (переполнения буфера) не должна превышать 2%. Что выгоднее экономически: увеличить буфер до нужного размера или добавить сервер?

**Решение**

В данной задаче имеем дело с многоканальной СМО с ограниченной очередью. В пакете queueing подходит модель NewInput.MMCK

In [18]:
lambda <-100
mu <- 40

c <- 3
m <-10

K <- c + m 

library(queueing)
input <- NewInput.MMCK(lambda = lambda, mu = mu, c = c, k = K)
model <- QueueingModel(input)

Warning message in formals(fun):
“аргумент -- не функция”
Warning message in formals(fun):
“аргумент -- не функция”
Warning message in formals(fun):
“аргумент -- не функция”


In [21]:
P_refuse <- model$Pn[K+1]  # вероятность отказа (состояние K)
cat("Вероятность отказа:", round(P_refuse*100, 4), "%\n")

Lq <- model$Lq  # средняя длина очереди
cat("Средняя длина очереди:", round(Lq, 2), "задач\n")

Wq <- model$Wq  # среднее время ожидания
cat("Среднее время ожидания:", round(Wq, 2), "мин\n")

Вероятность отказа: 2.0876 %
Средняя длина очереди: 2.21 задач
Среднее время ожидания: 0.02 мин


In [23]:
A <- lambda * (1 - P_refuse)  # абсолютная пропускная способность (задач/мин)

income_per_day <- A * 60 * 24 * 5  # 5 руб за задачу
cost_per_day <- c * 2000 * 24       # 2000 руб/час за сервер
profit <- income_per_day - cost_per_day

cat("Доход в сутки:", round(income_per_day/1000, 2), "тыс.руб\n")
cat("Расход в сутки:", round(cost_per_day/1000, 2), "тыс.руб\n")
cat("Чистая прибыль:", round(profit/1000, 2), "тыс.руб\n")

Доход в сутки: 704.97 тыс.руб
Расход в сутки: 144 тыс.руб
Чистая прибыль: 560.97 тыс.руб


Пробуем увеличить емкость системы:

In [24]:
lambda <-100
mu <- 40

c <- 3
m <-11

K <- c + m 

library(queueing)
input <- NewInput.MMCK(lambda = lambda, mu = mu, c = c, k = K)
model <- QueueingModel(input)

P_refuse <- model$Pn[K+1]  # вероятность отказа (состояние K)
cat("Вероятность отказа:", round(P_refuse*100, 4), "%\n")

Lq <- model$Lq  # средняя длина очереди
cat("Средняя длина очереди:", round(Lq, 2), "задач\n")

Wq <- model$Wq  # среднее время ожидания
cat("Среднее время ожидания:", round(Wq, 2), "мин\n")

A <- lambda * (1 - P_refuse)  # абсолютная пропускная способность (задач/мин)

income_per_day <- A * 60 * 24 * 5  # 5 руб за задачу
cost_per_day <- c * 2000 * 24       # 2000 руб/час за сервер
profit <- income_per_day - cost_per_day

cat("Доход в сутки:", round(income_per_day/1000, 2), "тыс.руб\n")
cat("Расход в сутки:", round(cost_per_day/1000, 2), "тыс.руб\n")
cat("Чистая прибыль:", round(profit/1000, 2), "тыс.руб\n")

Warning message in formals(fun):
“аргумент -- не функция”
Warning message in formals(fun):
“аргумент -- не функция”
Warning message in formals(fun):
“аргумент -- не функция”


Вероятность отказа: 1.7099 %
Средняя длина очереди: 2.36 задач
Среднее время ожидания: 0.02 мин
Доход в сутки: 707.69 тыс.руб
Расход в сутки: 144 тыс.руб
Чистая прибыль: 563.69 тыс.руб


Пробуем увеличить количество серверов:

In [25]:
lambda <-100
mu <- 40

c <- 4
m <-10

K <- c + m 

library(queueing)
input <- NewInput.MMCK(lambda = lambda, mu = mu, c = c, k = K)
model <- QueueingModel(input)

P_refuse <- model$Pn[K+1]  # вероятность отказа (состояние K)
cat("Вероятность отказа:", round(P_refuse*100, 4), "%\n")

Lq <- model$Lq  # средняя длина очереди
cat("Средняя длина очереди:", round(Lq, 2), "задач\n")

Wq <- model$Wq  # среднее время ожидания
cat("Среднее время ожидания:", round(Wq, 2), "мин\n")

A <- lambda * (1 - P_refuse)  # абсолютная пропускная способность (задач/мин)

income_per_day <- A * 60 * 24 * 5  # 5 руб за задачу
cost_per_day <- c * 2000 * 24       # 2000 руб/час за сервер
profit <- income_per_day - cost_per_day

cat("Доход в сутки:", round(income_per_day/1000, 2), "тыс.руб\n")
cat("Расход в сутки:", round(cost_per_day/1000, 2), "тыс.руб\n")
cat("Чистая прибыль:", round(profit/1000, 2), "тыс.руб\n")

Warning message in formals(fun):
“аргумент -- не функция”
Warning message in formals(fun):
“аргумент -- не функция”
Warning message in formals(fun):
“аргумент -- не функция”


Вероятность отказа: 0.1093 %
Средняя длина очереди: 0.51 задач
Среднее время ожидания: 0.01 мин
Доход в сутки: 719.21 тыс.руб
Расход в сутки: 192 тыс.руб
Чистая прибыль: 527.21 тыс.руб


Таким образом, выгоднее увеличить буфер до нужного размера.

## $\S 8.$ **Задачи для самостоятельного решения**

**Задача 1.**

На вход одноканальной СМО с отказами поступает простейший поток заявок с интенсивностью $\lambda=1$. Время обслуживания  - показательное с параметром $\mu=3$. В начальный момент времени $t=0$ канал свободен. Записать и решить диференциальные уравнения Колмогорова для вероятностей состояний СМО. Найти финальные вероятности и характеристики эффективности.

**Задача 2.**

Имеется одноканальная СМО с отказами. Поток с интенсивностью $\lambda$. Время обслуживания неслучайное и в точности равно $t_{\text{обсл}}=\frac{1}{\mu}$. Найти финальные вероятности состояний системы.

**Задача 3.**

Имеется двухканальная СМО с тремя местами в очереди при $\lambda=0.6, \mu=0.2$. Определить для этой СМО финальные вероятности и характеристики эффективности.

**Задача 4.**

Прибор состоит из двух узлов. Интенсивности отказов этих узлов $\lambda=0.02$, интенсивность их восстановления $\mu=1.0$. Определить вероятности состояний прибора. 

**Задача 5.** 

Имеется двухканальная простейшая СМО с отказами. На ее вход поступает поток заявок с интенсивностью 4 заявки/ч. Среднее время обслуживания одной заявки 0.8 часов. Каждая обслуженная заявка приносит доход 4 тыс.руб. Содержание каждого канала обходится в 2 тыс.руб/ч. Выяснить, выгодно ли увеличивать число каналов до трех. 